app/models/concerns/emits_entity_events.rb

```diff
@@ -0,0 +1,41 @@
+module EmitsEntityEvents
+  extend ActiveSupport::Concern
+
+  included do
+    attr_accessor :event_tags
+
+    after_create do
+      data = entity_data(attributes)
+      send_event("create_entity", data) if data.any?
+    end
+
+    after_update do
+      # in this after_update hook we don’t have access to the new fields via
+      # #attributes — we need to dig them out of saved_changes which stores
+      # them in the format { attr: ['old', 'new'] }
+
+      interesting_changes = entity_data(saved_changes.transform_values(&:last))
+
+      if interesting_changes.any?
+        send_event("update_entity", entity_data(attributes).merge(interesting_changes))
+      end
+    end
+  end
+
+  def send_event(type, data)
+    return unless FeatureService.enabled?(:send_request_data_to_bigquery)
+
+    event = BigQuery::EntityEvent.new do |ee|
+      ee.with_type(type)
+      ee.with_entity_table_name(self.class.table_name)
+      ee.with_data(data)
```

Comment:
Where does the request ID get set? I think we saw that Apply wasn't setting it so if this is copied from there ...

Reply:
We discussed the challenge around this briefly. Publish uses [request_store](https://github.com/steveklabnik/request_store) to do this ([setting](https://github.com/DFE-Digital/publish-teacher-training/blob/42ab166f600a51a45ce8506faf42c080e0213ae5/app/controllers/application_controller.rb#L291), [retrieving](https://github.com/DFE-Digital/publish-teacher-training/blob/e7a7a1a420f885f6645f9ea7b50906a4c7a94fa7/app/models/base.rb#L15)). Which reminds me, we should probably be setting this in the headers of API calls to Apply and PAPI too.

Reply:
Should we do that as another PR so we can start getting the events in?

Reply:
If you prefer ... so long as it doesn't get punted, I'd say.